src/Node/Internal.hs

```diff
@@ -383,31 +494,48 @@ nodeDispatcher endpoint nodeState handlerIn handlerInOut =
       case event of
 
           NT.ConnectionOpened connid NT.ReliableOrdered peer ->
-              -- Just keep track of the new connection, nothing else
-              loop (Map.insert connid (ConnectionNew (NodeId peer)) state)
+              -- Just keep track of the new connection and reverse lookup
+              -- from EndPointAddress.
+              loop $ state {
+                    dispatcherConnections = Map.insert connid (peer, ConnectionNew) (dispatcherConnections state)
+                  , dispatcherPeers = Map.alter (Just . maybe (NESet.singleton connid) (NESet.insert connid)) peer (dispatcherPeers state)
+                  }
 
           -- receiving data for an existing connection (ie a multi-chunk message)
           NT.Received connid chunks ->
-              case Map.lookup connid state of
-                  Nothing ->
-                      throw (InternalError "received data on unknown connection")
+              case Map.lookup connid (dispatcherConnections state) of
+                  Nothing -> do
+                      logDebug "internal error: received data on unknown connection"
+                      loop state
 
                   -- TODO: need policy here on queue size
-                  Just (ConnectionNew peer) ->
-                      loop (Map.insert connid (ConnectionNewChunks peer chunks) state)
-
-                  Just (ConnectionNewChunks peer@(NodeId _) chunks0) -> do
-                      mConnState <- handleFirstChunks connid peer (chunks0 ++ chunks) state
-                      loop (maybe state (\cs -> Map.insert connid cs state) mConnState)
+                  Just (peer, ConnectionNew) ->
+                      loop $ state {
+                            dispatcherConnections = Map.insert connid (peer, ConnectionNewChunks chunks) (dispatcherConnections state)
+                          }
+
+                  Just (peer, ConnectionNewChunks chunks0) -> do
+                      emConnState <- try $ handleFirstChunks connid peer (chunks0 ++ chunks) state
```

Comment:
Yes I think this is what we want. If there's an exception then we really do want to know about it.